analyzers/src/SonarAnalyzer.CSharp/Rules/ClassShouldNotBeEmpty.cs

```diff
@@ -0,0 +1,37 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ClassShouldNotBeEmpty : ClassShouldNotBeEmptyBase<SyntaxKind>
+{
+    protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
+
+    protected override bool IsEmptyClass(SyntaxNode node) => node is ClassDeclarationSyntax { Members.Count : 0 };
```

Comment:
Sounds like an interesting topic, but I'd say it should be addressed in a different PR. 
@andrei-epure-sonarsource what do you think?

Reply:
Hm, to avoid noise for source generators, maybe it's better to avoid raising this rule for partial classes?

For example, [this project](https://github.com/docopt/docopt.net/tree/master) is a source generator for creating argument parsers.

They use empty partial classes quite extensively 

- https://github.com/docopt/docopt.net/blob/master/src/DocoptNet/Public.cs
-  https://github.com/docopt/docopt.net/blob/master/src/DocoptNet/Internals/Public.cs
- https://github.com/docopt/docopt.net/blob/master/src/DocoptNet/CodeGeneration/Public.cs

I'd favor being on the safe side, avoid partial and eventually understand this coding pattern. We can make it more strict later on if we better understand how we can do that without introducing noise.

These are my 2c.

WDYT @zsolt-kolbay-sonarsource ?

Even if I don't fully understand the reasoning, I would still

Reply:
and to answer - if we know it can create noise, I'd change it in this PR

Reply:
Changed the analyzer to ignore partial classes and records entirely. Also updated the RSPEC.